### PR TITLE
Fix regression: open custom node and then open dyn mess up workspace

### PR DESCRIPTION
**Pull request by ke-yu**
This PR is to fix defect [MAGN-5919: regression: open custom node and then open dyn mess up workspace](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-5919).

The symptom is opening/closing .dyn and .dyf files few times, then some workspace stops responding to UI inputs. The cause is wrong workspace view model is used. 

Previously, switching to the other workspace model is done by calling
```
DynamoModel.SwitchTabCommand(int tabIndex)
```
The argument ``tabindex`` is the index of workspace view model in the workspace view model list that maintained by ``DynamoViewModel``: 
```
public int CurrentWorkspaceIndex
{
    ...
    set
    {
        var viewModel = workspaces.FirstOrDefault(vm => vm.Model == model.CurrentWorkspace);
        var index = workspaces.IndexOf(viewModel);
        ...
        this.ExecuteCommand(new DynamoModel.SwitchTabCommand(value));
    }
}
```
But, the order of workspace view model in the list could be different from the order of its corresponding workspace model in the model list that maintained by ``DyanmoModel``:
untitled

As ``DynamoModel.CurrentWorkspace`` could became invalid in the middle of some operation (for example, opening a new home workspace will remove the old home workspace and add a new one, and after removing operation, ``DynamoModel.CurrentWorkspace`` is invalid), or invalid because of order issue, its corresponding workspace view model could be invalid as well, therefore ``DynamoModel.CurrentWorkspaceIndex`` could be broken and cause mess in view). 

To fix this issue, firstly, we replace ``tabIndex`` (i.e., workspace view model index) in ``SwitchTabCommand()`` with ``modelIndex``, so that we could switch to correct workspace model.

Secondly,  ``currentWorkspaceViewModel`` is added to ``DynamoViewModel`` to keep track of current workspace view model, instead of coming from ``DynamoMode.CurrentWorkspace``. 

Issue: It is weird to have a ``SwitchTabCommand``in ``DynamoModel``, the latter shouldn't have any idea about UI, i.e., tab index. But removing it may cause recordable framework be broken, and @Benglin mentions this command is still needed in no UI mode. Any thought?

Need your comments :-) :
  - [ ] @Benglin 
  - [x] @pboyer

### CODE REVIEW

**Comment by pboyer** [comment]
Hi @ke-yu,

I think the approach you're taking here makes sense.  This is a serious issue that we should immediately sort out.  (Hence, I've checked the box above)

However, you're right.  It seems to me, actually, that *`DynamoModel` should know nothing about a current workspace index at all.*  That's a UI layer construct.  IOW, there's no reason, other than the limitation of the UI, that a user have a single workspace open.  Imagine how wrong this would be if we have multiple windows!

File: src/DynamoCore/Models/RecordableCommands.cs [hunk 1]
```diff
@@ -1107,9 +1107,9 @@ public class SwitchTabCommand : RecordableCommand
         {
             #region Public Class Methods
 
-            public SwitchTabCommand(int tabIndex)
+            public SwitchTabCommand(int modelIndex)
```
Comment (Benglin) [inline src/DynamoCore/Models/RecordableCommands.cs]: Minor thing, but `workspaceModelIndex` would have made a better argument name.

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs [hunk 2]
```diff
@@ -151,16 +152,35 @@ public int CurrentWorkspaceIndex
...
                 var index = workspaces.IndexOf(viewModel);
+
+                // As the getter could aslo be triggered by the change of model,
+                // we need to update currentWorkspaceViewModel here. 
+                if (currentWorkspaceViewModel != viewModel)
+                    currentWorkspaceViewModel = viewModel;
+
                 return index;
             }
             set
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: I'm not 100% certain that `currentWorkspaceViewModel` is needed. From a design perspective, a `DynamoViewModel` cannot live without `DynamoModel`, therefore `DynamoViewModel.CurrentWorkspaceIndex` property should always be inferred by `DynamoModel.CurrentWorkspace` and be made a read-only property. The only notion of "current workspace" should live in `DynamoModel` and nowhere else. In another word, if one were to switch the current workspace, it has to go through `SwitchTabCommand` (that needs a better name, I know), there is no second way around it. Is this possible? I'd prefer to examine all the callers that set `DynamoViewModel.CurrentWorkspaceIndex`, and see if they can be updated to use `SwitchTabCommand` instead.

**Comment by Benglin** [comment]
Hey @pboyer, I do agree that having the notion of "current workspace" in `DynamoModel` is a little weird, but that is mainly because some code paths require "current workspace" to be known when executed. For example, "I want to create NodeAbc", that statement does not make sense unless we know in which `WorkspaceModel` the `NodeAbc` should be created in. One way to know is to maintain a "current workspace" in `DynamoModel`, another way would be for the caller of `CreateNodeAbc` to pass in the `WorkspaceModel`, but then again the said caller would need to know *which workspace to deal with*.

As highlighted above I would propose to continue to have the concept of "current workspace" in `DynamoModel`, but ensure that it is the only ground-truth from where we determine the `WorkspaceModel` to work with. In another word, the concept of "current workspace" should not exist anywhere else (e.g. `DynamoViewModel`) other than in `DynamoModel.CurrentWorkspace`. If `DynamoViewModel` does expose "current workspace" or "current workspace index" properties, they **must** derive their values from `DynamoModel.CurrentWorkspace`, and be made read-only.

Does that make sense?

File: src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs (hunk 2 above)
Comment (ke-yu) [inline src/DynamoCoreWpf/ViewModels/Core/DynamoViewModel.cs]: Not all switching workspace going through ``SwitchTabCommand``. One is adding home workspace model, which 1) deletes current home workspace model, and consequently delete current home workspace view model (and cause ``CurrentWorkspaceIndex`` is triggered as workspace view mode list is the data source of DynamoView 2) and add new home workspace and new home workspace view model. During this periods, at some points ``CurrentWorkspace`` hasn't be properly set or is in invalid state.

**Comment by ke-yu** [comment]
@Benglin , in non-UI mode, it totally makes sense. But as we are talking about view model / model, the context should be MVVM. IMHO, maintain a "current workspace view model" probably makes more sense because 1) creating a node starts from UI 2) the command will be handled by ``DynamoViewModel`` which contains a current workspace view model 3) from current workspace view model we can get a workspace model, and add newly created node to workspace model.

Now back to non-UI node. We tend to think that the command will be executed directly on model, but this breaks the MVVM layers. Can we create a special view model to do that?

**Comment by Benglin** [comment]
Hi @ke-yu, is there anything else we need to discuss about this? I know we had a round of discussion, but not sure if we do have clarity on how a `CreateNodeCommand` is deserialized and target the right `WorkspaceModel` (since `WorkspaceModel` has a rather random `GUID`, unless of course we change that).

**Comment by ke-yu** [comment]
@Benglin , I think you are talking about `SwitchTabCommand` instead of `CreateNodeCommand`, right :-) Right now `SwitchTabCommand` serializes the index of workspace model, because it is `DynamoModel` which finally executes the command.

**Comment by Benglin** [comment]
Oops, sorry, you're right. I was navigating through 18 PRs this morning :(

**Comment by pboyer** [comment]
@Benglin, on your comment above.  No, I disagree.  There should be no current workspace on the model layer.  In your example, the command can easily pass along a GUID identifying which workspace the node should be added to.

**Comment by ke-yu** [comment]
Current workspace is a UI concept, so I think @pboyer is correct, model layer shouldn't have that. It probably is OK to be there because we add a workspace GUID to all commands related to workspace model (we have done that for `UpdateModelValueCommand()` which needs to specify a target workspace model instead of using current workspace model) , but the bottom line is, the control shouldn't be inverted. 

If we look at the original implementation of `DynamoViewModel.CurrentWorkspaceIndex`:
```
public int CurrentWorkspaceIndex
{
    ...
    set
    {
        var viewModel = workspaces.FirstOrDefault(vm => vm.Model == model.CurrentWorkspace);
        ...
    }
}
```
obviously the UI is controlled by `DynamoModel.CurrentWorkspace`. It is the cause. 

In short, `DynamoModel.CurrentWorkspace` by all means should be set by `DynamoViewModel` and shouldn't have impact on UI. We probably should remove `CurrentWorkspace` property changed event handler from `DynamoViewModel`. And in the long term, all commands need a target workspace model GUID.

**Comment by ke-yu** [comment]
Enough said. I've created a [user story](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-6421) to fully clean up these remaining issues and merge this fix.
